localstack/aws/connect.py

```diff
@@ -0,0 +1,473 @@
+    def get_client(
+        self,
+        service_name: str,
+        region_name: Optional[str],
+        aws_access_key_id: Optional[str] = None,
+        aws_secret_access_key: Optional[str] = None,
+        aws_session_token: Optional[str] = None,
+        endpoint_url: str = None,
+        config: Config = None,
+    ):
+        raise NotImplementedError()
+
+    def _get_client_post_hook(self, client: BaseClient) -> BaseClient:
+        """
+        This is called after the client is created by Boto.
+
+        Any modifications to the client can be implemented here in subclasses
+        without affecting the caching mechanism.
+        """
+        return client
+
+    # TODO @cache here might result in a memory leak, as it keeps a reference to `self`
+    # We might need an alternative caching decorator with a weak ref to `self`
+    # Otherwise factories might never be garbage collected
+    @cache
+    def _get_client(
+        self,
+        service_name: str,
+        region_name: str,
+        use_ssl: bool,
+        verify: bool,
+        endpoint_url: str,
+        aws_access_key_id: str,
+        aws_secret_access_key: str,
+        aws_session_token: str,
+        config: Config,
+    ) -> BaseClient:
+        """
+        Returns a boto3 client with the given configuration, and the hooks added by `_get_client_post_hook`.
+        This is a cached call, so modifications to the used client will affect others.
+        Please use another instance of the factory, should you want to modify clients.
+        Client creation is behind a lock as it is not generally thread safe.
+
+        :param service_name: Service to build the client for, eg. `s3`
+        :param region_name: Name of the AWS region to be associated with the client
+            If set to None, loads from botocore session.
+        :param aws_access_key_id: Access key to use for the client.
+            If set to None, loads from botocore session.
+        :param aws_secret_access_key: Secret key to use for the client.
+            If set to None, loads from botocore session.
+        :param aws_session_token: Session token to use for the client.
+            Not being used if not set.
+        :param endpoint_url: Full endpoint URL to be used by the client.
+            Defaults to appropriate LocalStack endpoint.
+        :param config: Boto config for advanced use.
+        :return: Boto3 client.
+        """
+        with self._create_client_lock:
+            client = self._session.client(
+                service_name=service_name,
+                region_name=region_name,
+                use_ssl=use_ssl,
+                verify=verify,
+                endpoint_url=endpoint_url,
+                aws_access_key_id=aws_access_key_id,
+                aws_secret_access_key=aws_secret_access_key,
+                aws_session_token=aws_session_token,
+                config=config,
+            )
+
+        return self._get_client_post_hook(client)
```

Comment:
Should the post-hook maybe also be executed in the context of the lock?

Reply:
The client creation is not thread safe, but using the client / adding hooks is. The lock is there purely to prevent issues with the session, so having this outside the lock should not effect anything.